pkg/controller/networkpolicy/networkpolicy_controller.go

```diff
@@ -805,25 +915,36 @@ func (n *NetworkPolicyController) addService(obj interface{}) {
 	for group := range groupKeySet {
 		n.enqueueInternalGroup(group)
 	}
+
+	// Find all ACNPs that are impacted by this Service and reprocess them.
+	affectedACNPsByService := n.filterACNPsByService(service.Namespace, service.Name)
+	for _, acnp := range affectedACNPsByService {
+		n.reprocessACNP(acnp)
+	}
 }
 
-// updatePod retrieves all internal Groups which refers to this Service
-// and enqueues the group keys for further processing.
+// updateService retrieves all internal Groups which refers to this Service
+// and enqueues the group keys for further processing. And reprocess the
+// ClusterNetworkPolicy with `toService` referred to this Service.
 func (n *NetworkPolicyController) updateService(oldObj, curObj interface{}) {
 	defer n.heartbeat("updateService")
 	oldService := oldObj.(*v1.Service)
 	curService := curObj.(*v1.Service)
 	klog.V(2).Infof("Processing Service %s/%s UPDATE event, selectors: %v", curService.Namespace, curService.Name, curService.Spec.Selector)
-	// No need to trigger processing of groups if there is no change in the Service selectors.
-	if reflect.DeepEqual(oldService.Spec.Selector, curService.Spec.Selector) {
-		klog.V(4).Infof("No change in Service %s/%s. Skipping group evaluation.", curService.Namespace, curService.Name)
-		return
+	// Only trigger processing of groups if there are changes in the Service selectors.
+	if !reflect.DeepEqual(oldService.Spec.Selector, curService.Spec.Selector) {
```

Comment:
For `filterACNPsByService` we could use a similar condition? i.e. only enqueue the ACNP affected if the service selector is empty. Otherwise any updates in the Service will be picked up by the corresponding Endpoint and we don't need to enqueue twice here

Reply:
Good point! Done.